New: Add --fix-type option to CLI (fixes #10855)

nzakas opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[x] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

Implemented the `--fix-type` option as described in #10855. Also included a tool for easily updating the type of each rule based on the `rule-types.json` file.

As for approach, I added a `fixTypes` property to the `CLIEngine` configuration. I chose to do this rather than overwriting the `fix` option. I think `fixTypes` is a better implementation because it exposes this functionality to code editors whereas just overriding the `fix` option would not have done so.

**Is there anything you'd like reviewers to focus on?**

1. Did I properly implement the change?
2. Did I write all appropriate tests?
3. Review `rule-types.json` to see if you disagree with any of the assigned types.

**Note:** The first commit on this PR contains the default `rule-types.json` that I created based solely on matching `meta.docs.category`. The second commit (which I will add soon) is my suggested changes.

not-an-aardvark commented on lib/rules/no-extra-boolean-cast.js:
```diff
@@ -17,6 +17,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "problem",
```
I'd consider `no-extra-boolean-cast` to be a `suggestion` -- it seems like it improves the style of code but probably wouldn't be pointing out a bug.

nzakas replied:
I agree. Let's see what others think.

platinumazure replied:
Agreed, this should be a "suggestion".

kaicataldo replied:
Agreed.

nzakas replied:
Set to "suggestion"

not-an-aardvark commented on lib/rules/no-unused-vars.js:
```diff
@@ -18,6 +18,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I'd consider `no-unused-vars` to be a `problem`; a lot of errors that it reports are a bugs that would result from forgetting to use a particular variable.

nzakas replied:
I agree, let's see what others think.

platinumazure replied:
I agree this should probably be a "problem".

nzakas replied:
Setting to "problem"

not-an-aardvark commented on lib/rules/default-case.js:
```diff
@@ -12,6 +12,8 @@ const DEFAULT_COMMENT_PATTERN = /^no default$/i;
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I'd consider `default-case` to be a `problem`, since a missing default is often a bug.

nzakas replied:
I can go either way on this, let's see what others think.

kaicataldo replied:
I feel similarly to @nzakas on this - I really feel like it could go either way.

nzakas replied:
We currently have this listed under "Best Practices", so I think it makes sense to stick with "suggestion" for now.

not-an-aardvark commented on lib/rules/no-fallthrough.js:
```diff
@@ -55,6 +55,8 @@ function hasBlankLinesBetween(node, token) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I'd consider `no-fallthrough` to be a `problem`, since a fallthrough is often a bug.

nzakas replied:
Agreed.

platinumazure replied:
I'm not sure I agree, for similar reasons to those I had laid out in `default-case`: If the rule is allowing the user to declare intent, then this is a suggestion, not a problem.

That said, fallthrough is usually a higher-impact issue than lack of default, so I could be okay with "problem" here.

kaicataldo replied:
I agree that this should probably be the same as `default-case`, though I could see going either way on it.

nzakas replied:
Setting to "problem".

nzakas commented on lib/cli-engine.js:
```diff
@@ -429,6 +428,24 @@ class CLIEngine {
+
+            this.options.fix = lintResult => {
+                const rules = this.getRules();
+                const matches = fixTypes.has(rules.get(lintResult.ruleId).meta.type);
```
Ooh, nice catch! I'll update that.

nzakas commented on lib/cli-engine.js:
```diff
@@ -429,6 +428,24 @@ class CLIEngine {
+            const originalFix = this.options.fix;
+
+            this.options.fix = lintResult => {
+                const rules = this.getRules();
```
Ah thanks, I'll look into that.

nzakas replied:
So I moved `getRules()` out but forgot about how rules can be added later. Hmmm. That's a tricky one.

nzakas replied:
Okay, here's the approach I took: Create a `_rulesCache` property that is populate each time `executeOnText()` or `executeOnFiles()` is called. That way, it's called just once before doing the operation instead of each time a lint result needs to be checked. That should be much more efficient, and deal with new rules being added.

nzakas commented on lib/options.js:
```diff
@@ -97,6 +97,11 @@ module.exports = optionator({
+        {
+            option: "fix-type",
+            type: "Array",
+            description: "Specify the types of fixes to apply"
```
Yup, will add.

nzakas commented on lib/cli-engine.js:
```diff
@@ -429,6 +428,24 @@ class CLIEngine {
         }
+
+        // setup special filter for fixes
+        if (this.options.fixTypes && this.options.fixTypes.length > 0) {
```
Yeah, I went back and forth on this and wanted to see what everyone thought. I'll add a check and throw an error if unexpected types are included.

nzakas commented on docs/user-guide/command-line-interface.md:
```diff
@@ -233,6 +234,24 @@ getSomeText | eslint --stdin --fix-dry-run --format=json
+
+This option allows you to specify the type of fixes to apply when using either `--fix` or `--fix-dry-run`. The three types of fixes are:
+
+1. `problem` - fix potential errors in the code
```
Yeah, I also went back and forth on this, and believe that "problem" is the correct term. Some of the other terms we use:

* "error" is a severity level, so definitely can't use that
* We use "messages" in the output data structure as a generic catchall for everything, so we don't need to worry about that.
* As best I can tell, "problem" is only ever used in the default formatter output. My feeling is that this isn't too confusing, but if others feel it is, I'd suggest changing the term to "messages" to echo what is in the data structure.

platinumazure replied:
Should we change the stylish formatter? We could use "issue" there, for example, whereas "issue" doesn't work as well for the rule type (IMO).

(I'm aware "issue" also has meaning in GitHub, but I'm not worried about confusion between lint output and GitHub at this point.)

nzakas replied:
I wouldn't want to change the formatter as part of this PR, as I think that's a breaking change (we never know who is parsing the output). I'm open to changing it, though as I said, I also think it's fine if we don't.

kaicataldo replied:
`problem` feels better to me than `messages` because `messages` sounds all-encompassing - it's not immediately clear how `messages` differs from `suggestion` or `style`

nzakas commented on lib/rules/no-undef.js:
```diff
@@ -25,6 +25,8 @@ function hasTypeOfOperator(node) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I can go either way on this, so let's see if we can get some more opinions.

platinumazure replied:
I lean towards "problem" myself.

nzakas replied:
Setting to "problem"

nzakas commented on tools/rule-types.json:
```diff
@@ -0,0 +1,266 @@
+    "wrap-iife": "suggestion",
+    "wrap-regex": "style",
+    "yield-star-spacing": "suggestion",
+    "yoda": "suggestion"
```
Yeah, that's a good idea.

nzakas commented on Makefile.js:
```diff
@@ -738,6 +742,7 @@ target.gensite = function(prereleaseVersion) {
                 `title: ${title}`,
                 "layout: doc",
                 `edit_link: https://github.com/eslint/eslint/edit/master/${filePath}`,
+                ruleType,
```
This change adds a `ruleType` field in a rule page's frontmatter in case we want to expose that somehow on the page.

not-an-aardvark commented on lib/cli-engine.js:
```diff
@@ -591,6 +635,12 @@ class CLIEngine {
         } else {
+
+            // if there's a cache, populate it
+            if ("_rulesCache" in this) {
```
Will this map contain rules from plugins? There's an outstanding bug in https://github.com/eslint/eslint/issues/10838 where the rules in `CLIEngine` aren't available until the first time the `CLIEngine` instance runs on text, because it doesn't load plugins from a config until it meeds to use that config to lint a file.

This might not be a problem here if the config is fully resolved before the rules are obtained, but it would be good to double-check that this is the case.

nzakas replied:
Yes. This call to `getRules()` is executed every time `executeOnFiles()` is called just before the linting actually happens. At that point, all plugins have been loaded and any calls to `linter.defineRule()` have been called. Note that `"_rulesCache" in this` only checks for the existence of the property (as a shorthand for having valid `fixTypes`) and does not check the property's value.

platinumazure commented on lib/rules/no-array-constructor.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
I'm struggling with this one: It's a clear AST change, but I can't imagine any side effects when switching from the constructor to the array literal...

I lean towards "suggestion" here since the change also helps avoid an edge case (redefinition of `Array` global).

nzakas replied:
I can see this as a suggestion. We should probably be consistent with all of the "no" rules.

platinumazure commented on lib/rules/no-bitwise.js:
```diff
@@ -22,6 +22,8 @@ const BITWISE_OPERATORS = [
 
 module.exports = {
     meta: {
+        type: "style",
```
I think this should be a "suggestion".

platinumazure commented on lib/rules/no-console.js:
```diff
@@ -17,6 +17,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "problem",
```
I see that the meta.docs.category is set to "Possible Errors", but honestly I feel that this should be a "suggestion" (and maybe the category should be changed).

nzakas replied:
Agree

platinumazure commented on lib/rules/no-continue.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Since fixing errors reported by this rule would require refactoring the code, should this be a "suggestion"?

platinumazure commented on lib/rules/no-dupe-class-members.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
What happens when there are duplicate class members? I assume this should be a "problem".

nzakas replied:
Yeah, you're probably right. Let's see what others think.

platinumazure commented on lib/rules/no-duplicate-imports.js:
```diff
@@ -101,6 +101,8 @@ function handleExports(context, importsInFile, exportsInFile) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
What happens when there are duplicate imports? If a runtime error would occur, I think this should be a "problem".

platinumazure commented on lib/rules/no-ex-assign.js:
```diff
@@ -13,6 +13,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "problem",
```
I think this should be a "suggestion". (Not sure why the category is "Possible Errors", either.)

nzakas replied:
I'm still of the mind that this should be a problem, as it's unlikely that you intentionally meant to overwrite a reference to an exception that is thrown.

platinumazure commented on lib/rules/no-extra-parens.js:
```diff
@@ -574,15 +575,13 @@ module.exports = {
-                            firstLeftToken.range[1] === node.left.range[1] ||
-
-                            /*
+                            (firstLeftToken.range[1] === node.left.range[1] || /*
```
Was this change intentional? (The parentheses is probably fine but the comment should probably begin on the next line.)

kaicataldo replied:
Ironic, given the rule this is in 🤣

platinumazure commented on lib/rules/no-nested-ternary.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Since fixing errors reported by this rule would require refactoring the code (albeit said refactoring is tiny), should this be a "suggestion"?

nzakas replied:
Yeah, I'd tend to agree.

platinumazure commented on lib/rules/no-new-object.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
For similar reasons as `no-array-constructor`, should this be a "suggestion"?

platinumazure commented on lib/rules/no-obj-calls.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "problem",
```
What sort of bug are we warning users away from with this rule? (Wondering if this should be a "suggestion".)

nzakas replied:
We're warning against people doing things like `Math()` and `JSON()`, which would cause a runtime error. As such, I think it should be a problem.

platinumazure commented on lib/rules/no-octal.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Octal literals are forbidden in some of the language versions, right? (Seems this should be a "suggestion" at least.)

nzakas replied:
In strict mode (and modules) they are forbidden.

I agree, seems like suggestion is the right type.

platinumazure commented on lib/rules/no-prototype-builtins.js:
```diff
@@ -10,6 +10,8 @@
 
 module.exports = {
     meta: {
+        type: "problem",
```
What's the possible bug here? (Should this be "suggestion"?)

platinumazure commented on lib/rules/no-regex-spaces.js:
```diff
@@ -13,6 +13,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "problem",
```
Although there are readability issues when this rule is violated, I don't think we should assume there is a bug much/most of the time. (I think this should be a "suggestion".)

platinumazure commented on lib/rules/no-restricted-syntax.js:
```diff
@@ -10,6 +10,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Similar to no-restricted-properties, etc., users could be using this for actual issues. I think this should be a "suggestion" (it's not purely stylistic).

platinumazure commented on lib/rules/no-sparse-arrays.js:
```diff
@@ -10,6 +10,8 @@
 
 module.exports = {
     meta: {
+        type: "problem",
```
Not sure I understand why this is a probable bug. (Should this be "suggestion"?)

platinumazure commented on lib/rules/no-ternary.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Since fixing errors reported by this rule would require refactoring the code (albeit a tiny one), should this be a "suggestion"?

nzakas replied:
Agree

platinumazure commented on lib/rules/no-this-before-super.js:
```diff
@@ -36,6 +36,8 @@ function isConstructorFunction(node) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
Weird things happen if trying to use `this` or `super` before invoking `super()`, right? Should this be a "problem"?

platinumazure commented on lib/rules/no-unused-vars.js:
```diff
@@ -372,22 +374,18 @@ module.exports = {
-                (
-                    parent.type === "AssignmentExpression" &&
+                (// in RHS of an assignment for itself. e.g. `a = a + 1`
+                    ((
```
The indentation and extra parentheses are a little confusing here. (Was this change intentional? I see a comment was moved as well...)

platinumazure commented on lib/rules/prefer-object-spread.js:
```diff
@@ -212,15 +212,19 @@ function defineFixer(node, sourceCode) {
 
 module.exports = {
     meta: {
+        type: "style",
```
Since fixing errors reported by this rule would require changing the statement, should this be a "suggestion"?

platinumazure commented on lib/rules/rest-spread-spacing.js:
```diff
@@ -11,13 +11,17 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I think this should be a "style" type rule.

nzakas replied:
Totally agree. I thought I had caught all of the spacing rules.

platinumazure commented on lib/rules/sort-imports.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
sort-imports is "suggestion" but sort-keys is "style"-- feels inconsistent.

nzakas replied:
Seems like they both should be suggestion.

platinumazure commented on lib/rules/sort-keys.js:
```diff
@@ -73,12 +73,15 @@ const isValidOrders = {
 
 module.exports = {
     meta: {
+        type: "style",
```
sort-imports is "suggestion" but sort-keys is "style"-- feels inconsistent.

nzakas replied:
Yeah, this seems like it should be a suggestion. Let's see what others think.

platinumazure commented on lib/rules/yield-star-spacing.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I think this should be "style".

nzakas replied:
Agree

platinumazure commented on tests/lib/cli.js:
```diff
@@ -887,6 +928,35 @@ describe("cli", () => {
 
         });
 
+        it("should pass fixTypes to CLIEngine when --fix-type is passed", () => {
```
Is this a duplicate of the test starting on line 778?

nzakas replied:
It certainly appears that way. Nice catch! I'll remove it.

nzakas commented on lib/rules/camelcase.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
I'm still in the "style" camp on this one. Any change to resolve this issue requires changing the value of an AST node but does not change the structure of the program.

From an end-user point of view, I think I would expect a rule enforcing naming conventions to be listed as a style preference.

not-an-aardvark replied:
I'd also lean towards `"suggestion"` here because in the long term, I think it would be valuable to be consistent about the idea that style rules don't modify the AST.

For example, if we add a way to disable all stylistic rules in the future, then someone using Prettier would be able to use this to turn off the rules that Prettier will handle. However, if we have exceptions where rules like `camelcase` are marked as stylistic despite modifying the AST, then a Prettier user would need to manually keep track of those rules and reenable them.

nzakas replied:
Okay, seems like others are rallying around suggestion, so happy to go along.

One thing I want to make sure to call out: there is a difference between modifying the structure of the AST vs. modifying a value inside of an AST node. Changing the structure of the AST means changing the meaning of the program while modifying a value inside of an AST node means the meaning of the program does not change even though a symbol within the program does. 

If everyone is more comfortable with "style" meaning changing only those things that are not represented in the AST, I can go along with that.

nzakas commented on lib/rules/func-style.js:
```diff
@@ -10,6 +10,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
I definitely see the risk with this one, and as such, I'm leaning more towards "suggestion" for this one.

kaicataldo replied:
I think `suggestion` is the way to go here as well.

nzakas replied:
Ok, seems like suggestion is the consensus here.

nzakas commented on lib/rules/key-spacing.js:
```diff
@@ -360,10 +362,9 @@ module.exports = {
         function isKeyValueProperty(property) {
             return !(
-                property.method ||
+                (property.method ||
```
I didn't do it manually. I wonder if jscodeshift did this for some reason when I ran it on the files.

nzakas commented on lib/rules/key-spacing.js:
```diff
@@ -360,10 +362,9 @@ module.exports = {
                 property.shorthand ||
-                property.kind !== "init" ||
-                property.type !== "Property" // Could be "ExperimentalSpreadProperty" or "SpreadElement"
+                property.kind !== "init" || property.type !== "Property") // Could be "ExperimentalSpreadProperty" or "SpreadElement"
```
No, once again I think jscodeshift did this.

nzakas commented on lib/rules/max-classes-per-file.js:
```diff
@@ -15,18 +15,22 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
I could go either way on this. On the one hand, I think rules that place limits on how code is structured that cannot be automatically fixed could be considered a style, I would also have no objection to calling it a suggestion.

kaicataldo replied:
This feels similar to the complexity rule, which is currently a `suggestion`.

nzakas replied:
Ok, consensus seems to be suggestion.

nzakas commented on lib/rules/max-depth.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Same response as `max-classes-per-file`. I think we should be consistent with all of these "max" rules.

kaicataldo replied:
I think this is mostly true, though I think `max-len` is definitely a `style` type.

kaicataldo commented on lib/rules/capitalized-comments.js:
```diff
@@ -108,13 +108,17 @@ function createRegExpForIgnorePatterns(normalizedOptions) {
 
 module.exports = {
     meta: {
+        type: "style",
```
Even though comment tokens are not part of the AST, I can see a similar argument to https://github.com/eslint/eslint/pull/10912/files#r226037811 being made here, since it would change the value of the comment token.

kaicataldo replied:
Non-blocking - I'm unsure of what the right course of action is here....

nzakas replied:
Yeah, I think I agree with you.

kaicataldo commented on lib/rules/no-empty.js:
```diff
@@ -16,6 +16,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "problem",
```
Is there a reason this is a `problem` while `no-empty-function` and `no-empty-pattern` are not?

nzakas replied:
This was just done by mapping our current categories to the new rule types, no other reason. Seems like "suggestion" is probably the right approach for all of these.

kaicataldo commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -26,6 +26,8 @@ Here is the basic format of the source file for a rule:
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
In seeing all the discussion about what `type` each rule should be (and seeing how some of them don't necessarily match `docs.catetegory` as was initially noted [here](https://github.com/eslint/eslint/issues/10855#issuecomment-421090765)), it seems like we're heavily weighting what the fix for the error would entail when we're deciding these categories.

I wonder if we should make this more explicit and call it `fixType`? I like the idea of just having a `type` field, but I'm not sure it's clear by the name what criteria we're using when we decide on these categories.

I don't want to hold this up; just wanted to see what others think.

nzakas replied:
We talked about this initially in the TSC meeting when we accepted the proposal. Though discussion about what type of fixes are applied has helped inform the type of each rule, I don't see this as just describing the type of fix that is applied (which is already specified, to some degree, by `meta.fixable`). The intent of `meta.type` is to help us better determine the importance of the rule rather than the type of fix to be applied. While this info is being used solely for fixes right now, there are a lot of ways we can use this information going forward.

nzakas replied:
Reference: https://gitter.im/eslint/tsc-meetings?at=5bad404e01fb5f4ae13176fd

nzakas commented on lib/rules/consistent-this.js:
```diff
@@ -10,6 +10,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
I agree, seems like suggestion is a better fit.

nzakas commented on lib/rules/max-lines-per-function.js:
```diff
@@ -69,6 +69,8 @@ function getCommentLineNumbers(comments) {
 
 module.exports = {
     meta: {
+        type: "style",
```
I don't think the type of fix should determine the type of rule. Based on previous comments about "max" rules, though, it seems like this should be a suggestion.

nzakas commented on lib/rules/max-lines.js:
```diff
@@ -17,6 +17,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "style",
```
Agree

nzakas commented on lib/rules/max-nested-callbacks.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "style",
```
Agree

nzakas commented on lib/rules/max-params.js:
```diff
@@ -19,6 +19,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "style",
```
Agree

nzakas commented on lib/rules/max-statements.js:
```diff
@@ -19,6 +19,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "style",
```
Agree

nzakas commented:
I've updated the rule types based on feedback. Please take a look at the most recent commit and comment on `rule-types.json` if you have further feedback.

nzakas commented:
So, after synthesizing all of the feedback about rule types, it seems clear to me that the word "style" is confusing pretty much everyone. I think that I've tried to redefine "style" to make it more narrow than most people seem to use, and if we are already having trouble with it, it might not be the best choice.

## Updated Proposal

I think we can make the rule types less confusing if we change "style" to "layout". That would mean we have these three categories:

1. **problem** - flags code that could cause an error
2. **suggestion** - flags code that could be done in a better way
3. **layout** - flags code related *only* to whitespace changes in between AST nodes

The "layout" type would be more restrictive than the current "style", but I think doing so makes the intent clearer.

Thoughts?

platinumazure commented:
I like the intent of narrowing the focus of style changes and promoting some current "style" rules to suggestions.

That said: Should token presence/absence or location also fall under "layout", or should those be suggestions? (E.g., semi, semi-style, comma-dangle, comma-style) Or have I just reopened the previous can of worms?

not-an-aardvark commented:
I don't feel extremely strongly about it, but I still think a "non-AST fixes only" mode would be useful for the reasons described in https://github.com/eslint/eslint/pull/10912#discussion_r226037811.

Re. the distinction between changing the structure of the AST versus the value of the AST nodes, I'm not sure it's useful to draw a line between the two. For example, if I changed the name of a `Identifier` variable reference to refer to a different variable which is also defined in scope, then I would be significantly changing the semantics of a program even though I would only have changed the value of an AST node rather than the structure of the AST.

nzakas commented:
@not-an-aardvark yeah, this is what I'm trying to get towards: a setting that doesn't touch the AST at all and also isn't confused by the terminology. I think "layout" isn't the best word, but we're not using it anywhere else so I think it's unambiguous and we can define it exactly how we want.

not-an-aardvark commented:
I'm okay with using the term "layout" -- I was more trying to suggest that we should consider things like semicolons and parentheses as "layout" issues rather than suggestions since they don't change the AST.

nzakas commented:
We could certainly consider semicolons and parentheses as part of layout,
I'm not opposed to that and I think wouldn't be too confusing.

On Wed, Oct 24, 2018 at 10:14 AM Teddy Katz <notifications@github.com>
wrote:

> I'm okay with using the term "layout" -- I was more trying to suggest that
> we should consider things like semicolons and parentheses as "layout"
> issues rather than suggestions since they don't change the AST.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/10912#issuecomment-432747395>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACWktgxcYHT7HyLXlEA38kP-YkaPlCjks5uoJ_5gaJpZM4XE4bQ>
> .
>


-- 

______________________________
Nicholas C. Zakas
@slicknet

Author, Principles of Object-Oriented JavaScript <http://amzn.to/29Pmfrm>
Author, Understanding ECMAScript 6 <http://amzn.to/29K1mIy>

kaicataldo commented:
I agree with @not-an-aardvark about having a category that is "non-AST fixes only". It would allow ESLint to be used in conjunction with other tooling in the current greater JS ecosystem (specifically Prettier).

I think `"layout"` works, but I do wonder if that holds up if we start including semicolons and parentheses as part of the category. The only other term I can think of is `"style"`, but then we're back to where we started!

I would be fine with either `"layout"` or `"style"`.

nzakas commented:
I think the nice thing about "layout" is that it doesn't have a tightly-coupled meaning to code at this point, so we are free to define it however we want. "style" means a lot of things to a lot of different people, so that becomes a more difficult term to redefine in an ESLint-specific way.

I'm going to go with "layout" as meaning whitespace, semicolons, commas, and parentheses and update this PR. Then we can do a final review and see if everyone thinks it makes sense.

nzakas commented:
Okay, I've reviewed each rule and made adjustments based on the last proposal. I *really* like this as I think it greatly clarified which type of rule each rule should be.

Please leave comments on `rule-types.json` if you think there is a mistake (this will make it easier for us all to load this ever-growing PR to review).

not-an-aardvark commented on lib/cli-engine.js:
```diff
@@ -459,6 +474,30 @@ class CLIEngine {
         }
+
+        // setup special filter for fixes
+        if (this.options.fixTypes && this.options.fixTypes.length > 0) {
```
I might be misunderstanding this logic, but won't this still apply fixes if `originalFix` is `false`? The logic in this `if` statement checks if `originalFix` is a function, but if `originalFix` is not a function it seems to always apply fixes.

It might be better to do something like this:

```suggestion
        if (this.options.fix && this.options.fixTypes && this.options.fixTypes.length > 0) {
```

nzakas replied:
Oh wow, nice catch. `cli` errors if you use `--fix-type` without `--fix`, so that's probably why I missed it, but it should definitely be checked here, too. I'll update and add a test.

not-an-aardvark commented on lib/rules/array-callback-return.js:
```diff
@@ -141,6 +141,8 @@ function isCallbackOfArrayMethod(node) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
This seems like it should be a `problem` since it's intended to point out bugs.

nzakas replied:
Agree

not-an-aardvark commented on lib/rules/consistent-return.js:
```diff
@@ -53,6 +53,8 @@ function isClassConstructor(node) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I think this could also be `problem` since the goal is to catch forgotten `return` statements, but I don't feel strongly about it.

nzakas replied:
The goal is really just to have consistency that `return` either always returns a value or doesn't, it doesn't actually flag an error. Consistency seems like it would be a suggestion.

not-an-aardvark commented on lib/rules/constructor-super.js:
```diff
@@ -92,6 +92,8 @@ function isPossibleConstructor(node) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
This seems like it should be a `problem`; the patterns reported by this rule will generally throw an error at runtime.

nzakas replied:
Agreed

not-an-aardvark commented on lib/rules/no-const-assign.js:
```diff
@@ -13,6 +13,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
This seems like it should be a `problem`; the patterns reported by this rule will generally throw an error at runtime.

nzakas replied:
Agreed

nzakas commented:
@eslint/eslint-tsc sorry to ping everyone, but I'm having trouble keeping this PR updated because it touches so many files. I believe this is ready for a final review, so I'd appreciate people taking a look and letting me know what you think.

not-an-aardvark commented on lib/rules/no-empty-pattern.js:
```diff
@@ -10,6 +10,8 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I'd consider this to be a `problem` since reported problems are usually bugs.

nzakas replied:
Agreed

not-an-aardvark commented on lib/rules/no-extra-semi.js:
```diff
@@ -18,6 +18,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "layout",
```
I think this should be `suggestion` since it removes `EmptyStatement` nodes from the AST.

nzakas replied:
Agreed. This really probably should have been called "no-empty-statements", but alas.

not-an-aardvark commented on lib/rules/no-func-assign.js:
```diff
@@ -13,6 +13,8 @@ const astUtils = require("../util/ast-utils");
 
 module.exports = {
     meta: {
+        type: "problem",
```
I'd be fine with either `problem` or `suggestion` here, but I think it should be the same as `no-class-assign`. Currently, `no-class-assign` is `suggestion` and `no-func-assign` is `problem`.

nzakas replied:
I agree, these should be consistent. They also seem like both are problems.

not-an-aardvark commented on lib/rules/no-new-symbol.js:
```diff
@@ -11,6 +11,8 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
This seems like it should be a `problem`; the patterns reported by this rule will generally throw an error at runtime.

nzakas replied:
Agreed

not-an-aardvark commented on lib/rules/no-self-assign.js:
```diff
@@ -165,6 +165,8 @@ function eachSelfAssignment(left, right, props, report) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
This seems like it could be `problem` since it usually catches bugs, but I don't feel strongly about it.

nzakas replied:
I agree.

not-an-aardvark commented on lib/rules/no-self-compare.js:
```diff
@@ -12,6 +12,8 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
This seems like it should be `problem` since it usually catches bugs.

nzakas replied:
Agreed

not-an-aardvark commented on lib/rules/no-unmodified-loop-condition.js:
```diff
@@ -165,6 +165,8 @@ function updateModifiedFlag(conditions, modifiers) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
I think this could be `problem` since the goal is to catch bugs, although there are also some cases where it reports false positives.

nzakas replied:
Agreed.

not-an-aardvark commented on lib/rules/valid-jsdoc.js:
```diff
@@ -16,6 +16,8 @@ const doctrine = require("doctrine");
 
 module.exports = {
     meta: {
+        type: "problem",
```
This seems like it could be a `suggestion` since a lot of its constraints are stylistic.

nzakas replied:
I disagree here. The main goal of the rule is to ensure that parameters in the JSDoc are matched with actual parameters in the function declaration. While it's true there are some other more stylistic constraints built-in (probably incorrectly), I think the main goal is to identify a problem.

nzakas replied:
After thinking about this some more, I've convinced myself that this should be a suggestion because it has to do with a comment rather than code.

btmills commented on lib/rules/no-buffer-constructor.js:
```diff
@@ -10,13 +10,17 @@
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
The `Buffer` constructor is deprecated in all currently-supported versions of Node.js. Do you think that is sufficient to consider this a problem?

nzakas replied:
Good point, I agree.

btmills commented on lib/rules/no-use-before-define.js:
```diff
@@ -136,6 +136,8 @@ function isInInitializer(variable, reference) {
 
 module.exports = {
     meta: {
+        type: "suggestion",
```
If this rule only checked TDZ issues, I’d advocate for it to be a problem. Because it also checks hoisting, I could go either way. The TDZ messages are in the “fix this, your program is broken” category, and those lose significance because the hosting messages are appropriately considered suggestions.

nzakas replied:
Hmm, yeah, this is a bit sticky. I think we should probably go with the most severe warning that comes out of the rule, so that would make this a problem.

nzakas commented on lib/cli-engine.js:
```diff
@@ -459,6 +474,30 @@ class CLIEngine {
+                const rule = this._rulesCache.get(lintResult.ruleId);
+                const matches = rule.meta && fixTypes.has(rule.meta.type);
+
+                return matches && (typeof originalFix === "function" ? originalFix(lintResult) : true);
```
I'm not sure this would make much of a difference between `typeof` just does a value lookup, which is pretty fast. That said, I'll still make the suggested change in the spirit of keeping things DRY.

nzakas commented:
Okay, I've incorporated the latest round of feedback and also updated the new rule issue template to mention the rule types. As it turns out, we already have three categories that almost exactly match up to what we ended up with discussing it on this PR. :)

Assuming no new feedback comes in tonight, I'll merge this tomorrow.

Thanks, everyone, for the feedback. I know this is a monster of a pull request and I appreciate the level of detail everyone has given it.

nzakas commented:
I just rebased to fix a conflict and the world exploded. I have no idea what could have changed that would cause so many tests to start failing. Anyone have any ideas? I'm completely stumped.

not-an-aardvark commented:
I think 802e9262b5cabc0872afd654ce1c9d0dc331f76b and 5525eb6aa86266352064cb7d5b8acdc3509c8b31 would have caused conflicts in `CLIEngine`.

not-an-aardvark commented:
Actually, looking at the test failures it seems like the build might be broken on master -- also see https://github.com/eslint/espree/issues/398.

not-an-aardvark commented:
If you rebase onto master the tests should be fixed.

nzakas commented:
Thanks @not-an-aardvark!
